specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,122 @@
+# Keccak256 For Variable Length Input
+
+We describe the circuit that can process input of arbitrary size.
+
+## Plain Keccak behavior
+
+Keccak256 use a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
+
+If the last part of the inputs is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
+
+## Circuit behavior
+
+### Lookup Region
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+#### Checks
+
+1. `hash_id` is sequential `next.id - curr.id in [0, 1]`
```

Comment:
I'm not sure if this is exactly needed here. This sounds more like a way to check that permutations are from the same hash operation. 

But that's already guaranteed by the accomulated length counters.

Reply:
We know this already due to the condition `input_len == acc_len` no?

We also know when to change when `acc_len_prev != acc_len_cur`

Reply:
Not really, `acc_len` is bumped 136 each row. So it wouldn't exactly equal to `input_len`.

Reply:
Ohh, so `acc_len` is for permutation, not for all of the input bytes? I thought we had a variable that was couting the length for all of the bytes.
If it's not the case, yes, that's a way to do it with the id :)